### PR TITLE
2. fix(perf): When writing blocks to disk, don't block other async tasks

### LINKED ISSUE #4155: Full synchronization test times out in CI
## Motivation

The full chain synchronization test is currently timing out in CI. We should find a way to speed up synchronization so that it doesn't time out.

### LINKED ISSUE #4779: Move zebra-state writes to blocking threads
## Motivation

`zebra-state` is taking a long time to finish writing blocks, and sometimes hangs completely.
This most often happens in CI.

You can see a bunch of hangs and really slow syncs in this run from PR #4776, even during checkpoint syncing:
https://github.com/ZcashFoundation/zebra/runs/7330216995?check_suite_focus=true#step:6:404

## Tasks

- See if PR #4757 helps
- Move database writes to a `tokio::spawn_blocking()` thread
- Check for CPU-intensive cryptography in database writes, and move it to a `rayon::scope_fifo()` thread (inside a `tokio::spawn_blocking()` thread)

Teor has a draft PR for the `tokio::spawn_blocking()` part (#4199), they just need to update it and finish it off.

### LINKED ISSUE #4790: Move note commitment tree updates to a dedicated blocking and CPU-heavy thread
## Motivation

Zebra's sync can be slow or fail, because note commitment tree updates involve some CPU-heavy cryptography.

So we want to move note commitment tree updates into a `rayon` CPU thread pool `scope`.

### API Reference

Blocking threads:
https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html

CPU-heavy threads:
https://docs.rs/rayon/latest/rayon/fn.scope_fifo.html

### Detailed Analysis

See https://github.com/ZcashFoundation/zebra/issues/4583#issuecomment-1186662862

### Designs

- [ ] Wrap the entire function that calls `append()`/`root()` in `rayon`, to minimise the cost of scheduling `tokio` and `rayon` threads
- [ ] We'd like to update sprout, sapling, and orchard note commitment trees in parallel, using Rayon's `join()`
- [ ] If there isn't a `tokio::spawn_blocking()` wrapping the update future, add it (see ticket #4779 and its PR)

Example code for `rayon` in `tokio` blocking threads:
https://github.com/ZcashFoundation/zebra/blob/81727d7df0ee57e6d4f2172c287e52e414c4e6a0/zebra-consensus/src/primitives/ed25519.rs#L121-L134

## Related Work

We'll move state updates to blocking tokio threads as part of #4779

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

We want to improve Zebra's sync speed to avoid test timeouts in CI (#4155).

One way to speed up the sync is move blocking tasks to dedicated threads, so they don't stop other async tasks on the same thread from running. This PR moves disk writes to a `tokio` blocking thread.
Close #4779

Partially addresses #2188

Depends-On: #4792

### Specifications

https://docs.rs/tokio/latest/tokio/task/fn.block_in_place.html

## Solution

- Move CommitBlock and CommitFinalizedBlock requests to a `tokio` block-in-place thread

Related changes:
- Remove mempool downloader requirement for a populated state

Test fixes:
- Add blocking tokio threads to tests that now need them
    - Remove `tokio::time::pause()` from tests that don't need it (it doesn't work with blocking threads)
- In tests that need `tokio::time::pause()`, stop writing blocks to the state
    - Stop requiring a populated state in the mempool downloader
- Improve failure debug output in some tests
- Improve reliability of some tests with multithreaded executors

## Review

Anyone can review this PR, it's urgent because CI is failing, and it blocks a release.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

- #4788

### CODE REVIEW

**Comment by conradoplg** [comment]
Running full sync test at https://github.com/ZcashFoundation/zebra/actions/runs/2227577101

**Comment by conradoplg** [comment]
Failed due to https://github.com/ZcashFoundation/zebra/issues/3991, cherry-picked fix https://github.com/ZcashFoundation/zebra/pull/4198, and restarted at:
https://github.com/ZcashFoundation/zebra/actions/runs/2228129142

**Comment by teor2345** [comment]
@mergifyio update

**Comment by teor2345** [comment]
Merged in CI fix #4206 from the main branch, and started a full sync:
https://github.com/ZcashFoundation/zebra/actions/runs/2229899691

The time to beat is about 5h47m.

**Comment by teor2345** [comment]
> The time to beat is about 5h47m.

It's about 5h43m, which is better, but not by much.

**Review by oxarbitrage — DISMISSED** [review]
It looks this will bring little benefit but no harm so should be ok to include it.

**Comment by teor2345** [comment]
This didn't happen in my local tests, or in the PR tests, only in Mergify:
> ---- components::inbound::tests::fake_peer_set::mempool_advertise_transaction_ids stdout ----
> The application panicked (crashed).
> Message:  internal error: entered unreachable code: `MempoolTransactionIds` requests should always respond `Ok(Vec<UnminedTxId>)`
> Location: zebrad/src/components/inbound/tests/fake_peer_set.rs:280

https://github.com/ZcashFoundation/zebra/runs/6217533371?check_suite_focus=true#step:3:994

**Comment by teor2345** [comment]
@mergifyio update

File: zebrad/src/components/inbound/tests/fake_peer_set.rs [hunk 1]
```diff
@@ -208,7 +208,7 @@ async fn mempool_push_transaction() -> Result<(), crate::BoxError> {
     Ok(())
 }
 
-#[tokio::test]
+#[tokio::test(flavor = "multi_thread")]
 async fn mempool_advertise_transaction_ids() -> Result<(), crate::BoxError> {
```
Comment (teor2345) [inline zebrad/src/components/inbound/tests/fake_peer_set.rs:218]: This test is failing intermittently, possibly due to a race condition.

**Comment by teor2345** [comment]
I'm going to try to quickly fix this on Monday, but I might just have to close the PR for now, and add this branch to ticket #2188 so we can fix it later.

**Comment by teor2345** [comment]
I don't have time to fix the fragile tests right now, so I'm going to leave these changes for later.

**Comment by teor2345** [comment]
The `mempool_push_transaction` and `mempool_advertise_transaction_ids` tests are unstable in a multi-threaded `tokio` runtime. They fail occasionally with the wrong inbound service response.

I've pushed a fix to try and wait for the mempool, I'm also testing the fix locally in a loop.

**Comment by teor2345** [comment]
> The `mempool_push_transaction` and `mempool_advertise_transaction_ids` tests are unstable in a multi-threaded `tokio` runtime. They fail occasionally with the wrong inbound service response.
> 
> I've pushed a fix to try and wait for the mempool, I'm also testing the fix locally in a loop.

These tests are stable now, they passed over 3000 repeated test runs.
(They used to fail around 1/50 times.)

File: zebra-state/src/service/non_finalized_state.rs [hunk 2]
```diff
@@ -238,12 +242,65 @@ impl NonFinalizedState {
...
+            // so clone the inner chain if needed, then add the new block.
+            //
+            // Pushing a block onto a Chain can launch additional parallel batches.
+            // TODO: should we pass _scope into Chain::push()?
+            scope.spawn_fifo(|_scope| {
+                let new_chain = Arc::try_unwrap(new_chain)
+                    .unwrap_or_else(|shared_chain| (*shared_chain).clone());
+                chain_push_result = Some(new_chain.push(contextual).map(Arc::new));
+            });
+        });
```
Comment (teor2345) [inline zebra-state/src/service/non_finalized_state.rs]: This is how I run validation and `Chain` updates in parallel in the non-finalized state.

I didn't split up sprout, sapling, and orchard note commitment trees into separate threads. I can do that in another PR.

File: zebra-state/src/service/finalized_state/zebra_db/shielded.rs [hunk 3]
```diff
@@ -231,34 +234,131 @@ impl DiskWriteBatch {
...
+
+            if !orchard_note_commitments.is_empty() {
+                scope.spawn_fifo(|_scope| {
+                    orchard_result = Some(Self::update_orchard_note_commitment_tree(
+                        orchard,
+                        orchard_note_commitments,
+                    ));
+                });
+            }
+        });
```
Comment (teor2345) [inline zebra-state/src/service/finalized_state/zebra_db/shielded.rs]: This is how I run sprout, sapling, and orchard note commitment tree updates in parallel in the finalized state.

File: zebra-state/src/service/finalized_state/zebra_db/shielded.rs [hunk 4]
```diff
@@ -329,6 +428,41 @@ impl DiskWriteBatch {
...
+            });
+
+            scope.spawn_fifo(|_scope| {
+                sapling_root = Some(sapling.root());
+            });
+
+            scope.spawn_fifo(|_scope| {
+                orchard_root = Some(orchard.root());
+            });
+        });
```
Comment (teor2345) [inline zebra-state/src/service/finalized_state/zebra_db/shielded.rs]: This is how I run sprout, sapling, and orchard note commitment tree root calculations in parallel in the finalized state.

We could remove this code, and just calculate the roots after the note commitment tree update. (The root is cached.)

What do you think?

File: zebra-state/src/service.rs [hunk 5]
```diff
@@ -779,7 +779,19 @@ impl Service<Request> for StateService {
...
+                // Allow other async tasks to make progress while blocks are being verified
+                // and written to disk. But wait for the blocks to finish committing,
+                // so that `StateService` multi-block queries always observe a consistent state.
+                //
+                // Since each block is spawned into its own task,
+                // there shouldn't be any other code running in the same task,
+                // so we don't need to worry about blocking it:
+                // https://docs.rs/tokio/latest/tokio/task/fn.block_in_place.html#
+                let rsp_rx =
+                    tokio::task::block_in_place(|| self.queue_and_commit_non_finalized(prepared));
```
Comment (teor2345) [inline zebra-state/src/service.rs:794]: This is how I move database writes and waiting on `rayon` threads out of `tokio` futures, for the non-finalized state.

`block_in_place` still blocks any other code running concurrently in the same task. But we spawn a separate task for each block download and verify, and each checkpoint block commit, so we should be ok here.

File: zebra-state/src/service.rs [hunk 6]
```diff
@@ -804,7 +816,16 @@ impl Service<Request> for StateService {
...
+
+                // # Performance
+                //
+                // Allow other async tasks to make progress while blocks are being verified
+                // and written to disk. But wait for the blocks to finish committing,
+                // so that `StateService` multi-block queries always observe a consistent state.
+                //
+                // See the note in `CommitBlock` for more details.
+                let rsp_rx =
+                    tokio::task::block_in_place(|| self.queue_and_commit_finalized(finalized));
```
Comment (teor2345) [inline zebra-state/src/service.rs:828]: This is how I move database writes and waiting on `rayon` threads out of `tokio` futures, for the finalized state.

File: zebrad/src/components/inbound/tests/fake_peer_set.rs [hunk 7]
```diff
@@ -308,7 +316,7 @@ async fn mempool_advertise_transaction_ids() -> Result<(), crate::BoxError> {
     Ok(())
 }
 
-#[tokio::test]
+#[tokio::test(flavor = "multi_thread")]
 async fn mempool_transaction_expiration() -> Result<(), crate::BoxError> {
```
Comment (teor2345) [inline zebrad/src/components/inbound/tests/fake_peer_set.rs:320]: This test also needs the same fix for the second request:
>  Message:  internal error: entered unreachable code: `MempoolTransactionIds` requests should always respond `Ok(Vec<UnminedTxId>)`
> Location: zebrad/src/components/inbound/tests/fake_peer_set.rs:487

https://github.com/ZcashFoundation/zebra/runs/7423391225?check_suite_focus=true#step:3:6828

It's a bit further down and hidden in the diff:
https://github.com/ZcashFoundation/zebra/pull/4199/files#diff-6a7425f2d6ab0ef475e13cb6412322b72023e7139a23e04f0bc70205846d42b6R482-R490

File: zebra-state/src/service/check.rs [hunk 8]
```diff
@@ -104,27 +104,8 @@ where
 
 /// Check that the `prepared` block is contextually valid for `network`, using
```
Comment (conradoplg) [inline zebra-state/src/service/check.rs]:
```suggestion
/// Check that the block is contextually valid for `network`, using
```

File: zebra-state/src/service/finalized_state/zebra_db/shielded.rs (hunk 4 above)
Comment (conradoplg) [reply zebra-state/src/service/finalized_state/zebra_db/shielded.rs]: Not sure if I understand your question, the second approach would be to keep the original logic, or something different?

File: zebra-state/src/service/check/anchors.rs [hunk 9]
```diff
@@ -3,107 +3,84 @@
...
-                            .get(&joinsplit.anchor)
-                            .cloned()
-                            .or_else(|| {
-                                finalized_state
-                                    .sprout_note_commitment_tree_by_anchor(&joinsplit.anchor)
-                            })
-                    });
+) -> Result<HashMap<sprout::tree::Root, Arc<sprout::tree::NoteCommitmentTree>>, ValidateContextError>
+{
+    let mut sprout_final_treestates = HashMap::new();
```
Comment (conradoplg) [inline zebra-state/src/service/check/anchors.rs]: I need to review this more carefully but on a quick glance I was kind of confused why the function was renamed and documentation changed to remove Sprout but it still does Sprout things

**Review by conradoplg — COMMENTED** [review]
I'm halfway through the review and will probably resume it tomorrow, so apologies if some of the comments do not make sense, but I'll post them anyway

File: zebra-state/src/service/non_finalized_state.rs (hunk 2 above)
Comment (teor2345) [reply zebra-state/src/service/non_finalized_state.rs]: I moved this change to PR #4802, so @oxarbitrage can easily re-review this PR.

File: zebra-state/src/service/finalized_state/zebra_db/shielded.rs (hunk 3 above)
Comment (teor2345) [reply zebra-state/src/service/finalized_state/zebra_db/shielded.rs]: I moved this change to PR #4802, so @oxarbitrage can easily re-review this PR.

File: zebra-state/src/service/finalized_state/zebra_db/shielded.rs (hunk 4 above)
Comment (teor2345) [reply zebra-state/src/service/finalized_state/zebra_db/shielded.rs]: We could run `sprout.root()` after updating the `sprout` tree, in the same `rayon` task that updates the `sprout` tree.

I'll make this change in a separate commit in PR #4802.

File: zebrad/src/components/inbound/tests/fake_peer_set.rs (hunk 7 above)
Comment (teor2345) [reply zebrad/src/components/inbound/tests/fake_peer_set.rs:320]: I fixed this by updating the existing test fix commit.

File: zebra-state/src/service/check.rs (hunk 8 above)
Comment (teor2345) [reply zebra-state/src/service/check.rs]: I'll make this fix in PR #4802.

File: zebra-state/src/service/check/anchors.rs (hunk 9 above)
Comment (teor2345) [reply zebra-state/src/service/check/anchors.rs]: I'll split this function further in PR #4802.

File: zebra-state/src/service/finalized_state/zebra_db/shielded.rs (hunk 4 above)
Comment (teor2345) [reply zebra-state/src/service/finalized_state/zebra_db/shielded.rs]: Actually if we do this change, there's a risk we'll calculate the roots when we don't need to. (If we end up updating the trees multiple times.)

So I'll just add a TODO instead.

File: zebra-state/src/service/check.rs (hunk 8 above)
Comment (teor2345) [reply zebra-state/src/service/check.rs]: Fixed in b1987f495d

File: zebra-state/src/service/check/anchors.rs (hunk 9 above)
Comment (teor2345) [reply zebra-state/src/service/check/anchors.rs]: Split sprout into its own function in 5ecc9393b0

**Comment by teor2345** [comment]
I admin-merged this PR because it is an urgent sync fix, and it has already been reviewed.
(Except for some minor test fixes.)

**Pull request merged**